### DISCUSSION TITLE
Metric selection

### DISCUSSION THREAD

**Discussion by Itsisme** [jvde-github/AIS-catcher#84] [Ideas]
Hai,

Would a metric selection be an option ?

Rgds

**Comment by jvde-github** [comment]
Hi,

can you give a bit more detail on what you like to see?

Thanks! Jasper

**Reply by Itsisme** [reply]
Hai Jasper, 

It would be nice to see speed and distance in metrics (optional)

![AISc-dist2](https://user-images.githubusercontent.com/126970761/222917622-d59713d9-819e-401c-8726-a3adeabe62d6.jpg)
![AISc-dist1](https://user-images.githubusercontent.com/126970761/222917624-cf0112c7-e277-4c9e-81b3-a5293e6594fe.jpg)

Also in the stats selectable nmi or km (as a non boatsman NMi & kts does not say to much :) )

Great piece of software.

Rgds JP

**Comment by jvde-github** [comment]
I made it into an URL option, e.g. for standard units as per AIS specification:
```
http://raspberrypi:8100/?metric=DEFAULT
```
Meters, km/h etc:
```
http://raspberrypi:8100/?metric=SI
```
and mph, mi, ft: 
```
http://raspberrypi:8100/?metric=US
```
The plots need to wait for another day. Does this work for you? Hope I got the conversions right :-)

ps I know people are running plugins to do conversions, the above does not work if overwritten
ps2 the settings should be stored for the next run so only have to be provided once on the URL

**Reply by Itsisme** [reply]
Hai,

Seems to work ok.
In the graphic screens I still see NMi (various graphs) in the popup per ship I see KM now perfect.
The plots needs another day for updating if I understand so I will wait for that to see if it follows.

Many thxs...

**Comment by jvde-github** [comment]
Hi. Found some time already to also update the chart page. Can you check whether this works? Thanks!

**Reply by Itsisme** [reply]
Runs great bedankt..... now all in Metric......

ps I now attached two dongels (2mhz bandwith each) I have two instances running one with -c CD and one without can I run one instance does it recognize multiple SDR's tried a few things but nogo yet :) ?

**Comment by jvde-github** [comment]
Super, graag gedaan. 

Unfortunately, it is a single SDR program. But what you can do if you want one web interface for the two dongles: run 3 versions.
You can send the output of both your AIS-catcher runs to one UDP address, e.g.:
```
AIS-catcher [-c CD] -u 192.168.1.235 4004 JSON on
```
and then run a third to pick up these messages and act as the webserver or do other things:
```
AIS-catcher -x 192.168.1.235 4004 -N 8100 etc etc
```
Unfortunately, this is how it is for now. 

Hope this helps.

**Reply by Itsisme** [reply]
Hai,

Tried that sulution but weird enough the counts go down quite a bit doing it with the JSON solution.
Not sure why I get that effect.
For now left two instances and it seems to feed well.

Need to play with it a bit more I guess :)

Chrs..

**Reply by jvde-github** [reply]
Thanks, will also do some extra testing. Any error messages in the log?

**Reply by jvde-github** [reply]
Ok, for good order. I think this has been fixed (in the -b multiple branch)

**Comment by jvde-github** [comment]
I did some experimentation and got a version running that can take input from multiple devices in parallel. So, hopefully in the next few weeks we have a version that only needs one instance of the program running :-)

**Comment by Itsisme** [comment]
Ah that sounds good...... got enough HW.....(linux)

  0:  RTL-SDR KerberosSDR   SN 00000001  A VHF/UHF/SHF X7000N
  1:  RTL-SDR KerberosSDR   SN 00000002  B AIS
  2:  RTL-SDR KerberosSDR   SN 00000003  C HF bands HyEndFed 10/80m
  3:  RTL-SDR KerberosSDR   SN 00000004  D 50Mhz/VHF/UHF or V2000
  4:  Realtek RTL2838UHIDIR SN 00000001  E HF bands 80 meter long wire 1:49
  5:  Realtek RTL2832U      SN 00001236  ADSB
  6:  Realtek RTL2838UHIDIR SN 00000001  ACARS
  7:  Realtek RTL2838UHIDIR SN 00000101  Radiosonde HAB
  
      SDRplay RSP1 Ebay model            F FM Broadcast / Fun on discone
      FIFI SDR                           HF bands MLA-30+ loo

And still some on winblows.
So eager to test it for you... :)

**Comment by jvde-github** [comment]
Haha great, I tested it so far on the KrakenSDR with 5 input channels. So far so good but needs a bit more clean up.

Will let you know once it is ready, thanks for helping testing in advance :-)

```
AIS-catcher (build Mar  6 2023) v0.44-225-gff6892e
(C) Copyright 2021-2023 jvde-github and other contributors
This is free software; see the source for copying conditions.There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Searching for device with SN 1000.
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!
Searching for device with SN 1001.
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!
Searching for device with SN 1002.
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!
Searching for device with SN 1003.
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!
Searching for device with SN 1004.
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!
Device    : RTL2838UHIDIR
Settings  : rate 1536K format CU8 tuner AUTO rtlagc ON biastee OFF buffer_count 2
Model #0  : [AIS engine v0.44] ps_ema ON afc_wide OFF droop ON fp_ds OFF
[AIS engine v0.44 #0]                   received: 0 msgs, total: 0 msgs, rate: 0 msg/s
----------------------
[AIS engine v0.44 #0]                   total: 0 msgs
```

**Reply by Itsisme** [reply]
Nice 5 channel...expensive..  Could use another kerberos 4 channel :)

**Comment by Itsisme** [comment]
Is it possible to get a legenda on the colors and signs of the map ?

![afbeelding](https://user-images.githubusercontent.com/126970761/223501653-d20d9411-e377-4d48-8826-016f793d149f.png)

**Reply by jvde-github** [reply]
Sure, I will add it to the list but might take a bit of time.

**Reply by jvde-github** [reply]
Nice coverage you have btw. I recently added coloring on the basis of ERI codes, looks nice for inland shipping!

**Comment by jvde-github** [comment]
So, I have uploaded a version of the program that can handle multiple input devices. You can build it via the normal route (not for windows though) but clone it like this:
```
git clone https://github.com/jvde-github/AIS-catcher.git -b multiple
```

Then you can run with two devices like:
```
AIS-catcher -d serial1 -v -d serial2 -c CD -v -N 8100
```
Hope this works for you :-)

**Comment by Itsisme** [comment]
Oeps compiles... ok recognizes multiple SDR's but no map data and graphs stay empty but is seeing data:

(with one sdr but also with two sdr's) but nothing on display here...
[AIS engine v0.44 #0-0]                 received: 47 msgs, total: 146 msgs, rate: 15.6385 msg/s
[AIS engine v0.44 #0-0]                 received: 50 msgs, total: 196 msgs, rate: 16.6401 msg/s
[AIS engine v0.44 #0-0]                 received: 52 msgs, total: 248 msgs, rate: 17.3068 msg/s

Weird... old version runs fine..

**Reply by jvde-github** [reply]
Thanks for trying! Can you try to add `-M DTM` when running with one SDR? 
Btw quite some fundamental changes were required so can be that things have broken. 

On a different note, I am also testing the UDP passthrough performance and noticed with high data rates the UDP server is reading too slowly which might cause messages to be lost. I did not notice it but you seem to have a super high message rate.

Thanks for all your feedback.

**Reply by jvde-github** [reply]
The ship data for the map is in (assuming the webserver runs on raspberrypi): 
```
http://raspberrypi:8100/ships.json
```
This then likely is empty. To get all data that gas been captured but potentially not displayed as it is too old:
```
http://raspberrypi:8100/ships_full.json
```

**Reply by Itsisme** [reply]
With one SDR and -M DTM it works again....

**Comment by Itsisme** [comment]
Ps running on Ubuntu 22.04.2 LTS

**Reply by jvde-github** [reply]
You probably can then run with two devices like this:
```
AIS-catcher -d serial1 -M DTM -v -d serial2 -M DTM -v -N 8100
```
I noticed this issue when testing the new code and thought I solved it but apparently my fix is not in or not correct.

I also made some performance improvements to the UDP server so hopefully no more messages lost when passing through.

**Reply by jvde-github** [reply]
Ok, I uploaded a patch so I hope you can use again without `-M DTM`:
```
AIS-catcher -d serial1 -v -d serial2 -v -N 8100
```
Thanks for your help

**Reply by Itsisme** [reply]
Hai,
 
Just updated 'git pull' and started it without the -M DTM and looks good now many thnxs..

Some points (minor but nice :) )
- In summary screen add the SDR's (now showing only one but two running)
- total Vessel seen graph added
- Color legenda

Question what is the last part of the summary:

![afbeelding](https://user-images.githubusercontent.com/126970761/223843998-04db4149-be77-4443-bdab-e1e10424b9be.png)

What is that exact and how does it get that data?

Great job now already changed anonther station from aisdipatcher & rtl_ais to this..

Chrs..

**Reply by jvde-github** [reply]
Great, many thanks for testing and using it for your station.

The other items I have added to the list. 

The summary in the screen is driven by the counting the various message types. AIS has 27 different messages types (base station send msg 4, SAR, msg 9, AtoN msg 21, etc). These statistics are just the count of these.
Cheers, Jasper

**Comment by otm64** [comment]
Thanks for adding an option for multiple devices. I updated to latest version today and it works fine for me.
